File: src/pt/joaomneto/titancompanion/adventure/impl/AODAdventure.java
package pt.joaomneto.titancompanion.adventure.impl;

import android.os.Bundle;
import android.view.Menu;

import java.io.BufferedWriter;
import java.io.IOException;

import pt.joaomneto.titancompanion.R;
import pt.joaomneto.titancompanion.adventure.Adventure;
import pt.joaomneto.titancompanion.adventure.impl.fragments.aod.Army;

public class AODAdventure extends Adventure {


    public static final int FRAGMENT_SOLDIERS = 1;
    public static final int FRAGMENT_COMBAT = 2;
    public static final int FRAGMENT_EQUIPMENT = 3;
    public static final int FRAGMENT_NOTES = 4;

    private Army soldiers = new Army();


    public AODAdventure() {
        super();
        fragmentConfiguration.clear();
        fragmentConfiguration.put(FRAGMENT_VITAL_STATS, new AdventureFragmentRunner(R.string.vitalStats, "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureVitalStatsFragment"));
        fragmentConfiguration.put(FRAGMENT_SOLDIERS, new AdventureFragmentRunner(R.string.soldiers, "pt.joaomneto.titancompanion.adventure.impl.fragments.aod.AODAdventureSoldiersFragment"));
        fragmentConfiguration.put(FRAGMENT_COMBAT, new AdventureFragmentRunner(R.string.fights, "pt.joaomneto.titancompanion.adventure.impl.fragments.rp.RPAdventureCombatFragment"));
        fragmentConfiguration.put(FRAGMENT_EQUIPMENT, new AdventureFragmentRunner(R.string.goldTreasure, "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureEquipmentFragment"));
        fragmentConfiguration.put(FRAGMENT_NOTES, new AdventureFragmentRunner(R.string.notes, "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureNotesFragment"));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        try {
            super.onCreate(savedInstanceState);

        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.adventure, menu);
        return true;
    }

    @Override
    public void storeAdventureSpecificValuesInFile(BufferedWriter bw) throws IOException {

        bw.write("gold=" + getGold() + "\n");
        bw.write("soldiers=" + getSoldiers().getStringToSaveGame());
    }

    @Override
    protected void loadAdventureSpecificValuesFromFile() {
        setGold(Integer.valueOf(getSavedGame().getProperty("gold")));
        setSoldiers(Army.getInstanceFromSavedString(getSavedGame().getProperty("soldiers")));

    }

    public String getCurrencyName() {
        return getResources().getString(R.string.gold);
    }

    public Army getSoldiers() {
        return soldiers;
    }

    public void setSoldiers(Army soldiers) {
        this.soldiers = soldiers;
    }



}

File: src/main/java/pt/joaomneto/titancompanion/adventure/impl/AODAdventure.kt
package pt.joaomneto.titancompanion.adventure.impl

import android.view.Menu
import pt.joaomneto.titancompanion.R
import pt.joaomneto.titancompanion.adventure.Adventure
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureCombatFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureEquipmentFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureNotesFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureVitalStatsFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.aod.AODAdventureSoldiersFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.aod.Army
import pt.joaomneto.titancompanion.adventure.state.AdventureState
import pt.joaomneto.titancompanion.adventure.state.StateKey
import pt.joaomneto.titancompanion.util.AdventureFragmentRunner
import java.util.*

class AODAdventure : Adventure<AODState>(
        arrayOf(
                AdventureFragmentRunner(R.string.vitalStats, AdventureVitalStatsFragment::class),
                AdventureFragmentRunner(R.string.soldiers, AODAdventureSoldiersFragment::class),
                AdventureFragmentRunner(R.string.fights, AdventureCombatFragment::class),
                AdventureFragmentRunner(R.string.goldTreasure, AdventureEquipmentFragment::class),
                AdventureFragmentRunner(R.string.notes, AdventureNotesFragment::class)
        )
        , AODState::class
) {

    override val currencyName = R.string.gold

    val soldiersFragment: AODAdventureSoldiersFragment?
        get() = getFragment(AODAdventureSoldiersFragment::class)

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.adventure, menu)
        return true
    }

    override fun loadAdventureSpecificStateFromSavedGame(
            values: Map<out StateKey, Any>,
            savedGame: Properties
    ) = AODState(
            values,
            Army.getInstanceFromSavedString(savedGame.getProperty("soldiers"))
    )

    fun test() {
        state = state.increaseGold()
    }
}

data class AODState(
        override val values: Map<out StateKey, Any>
) : AdventureState(values) {

    val soldiers
        get() = (values[AODStateKey.SOLDIERS] as Army)

    constructor(
            state: Map<out StateKey, Any>,
            soldiers: Army
    ) : this(state.plus(AODStateKey.SOLDIERS to soldiers))

    override fun storeAdventureSpecificValuesInFile() = "soldiers=" + soldiers.stringToSaveGame
}

enum class AODStateKey(override val saveFileKey: String,
                       override val serializer: ((Any) -> String)? = null) : StateKey {
    SOLDIERS("soldiers", { (it as Army).stringToSaveGame })
}
